File: 202.happy-number.java
/*
 * @lc app=leetcode id=202 lang=java
 *
 * [202] Happy Number
 */

// @lc code=start
class Solution {
    public boolean isHappy(int n) {
        int slow = n;
        int fast = sumDigits(n);

        while (fast != slow) {
            slow = sumDigits(slow);
            fast = sumDigits(sumDigits(fast));
        }

        return fast == 1;
    }

    public int sumDigits(int n) {
        int sum = 0;

        while (n > 0) {
            int reminder = n % 10;
            sum += reminder * reminder;
            n /= 10; 
        }

        return sum;
    }
}
// @lc code=end

File: 202.happy-number.py
#
# @lc app=leetcode id=202 lang=python3
#
# [202] Happy Number
#

# @lc code=start
class Solution:
    def isHappy(self, n: int) -> bool:
        
        def sum_digits(num):
            digits_sum = 0
            while num > 0:
                digits_sum += pow(num % 10, 2)
                num //= 10

            return digits_sum
         
        
        slow, quick = n, sum_digits(n)
        
        # If the quick pointer detects 1, end the loop in advance
        while quick != 1 and slow != quick:
            slow = sum_digits(slow)
            quick = sum_digits(sum_digits(quick))
        return quick == 1

        

# @lc code=end

